### PR TITLE
fix: check willretain and willqos when WillFlag set to `true`

### PR DESCRIPTION
**Pull request by JimMoen**
Fixes [EMQX-12503](https://emqx.atlassian.net/browse/EMQX-12503)

Release version: v/e5.7.2

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [x] Added property-based tests for code which performs user input validation
- [x] Changed lines covered in coverage report
- ~[ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files~
- [x] For internal contributor: there is a jira ticket to track this change
- ~[ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket~
- ~[ ] Schema changes are backward compatible~

## Checklist for CI (.github/workflows) changes

- ~[ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)~
- ~[ ] Change log has been added to `changes/` dir for user-facing artifacts update~

[EMQX-12503]: https://emqx.atlassian.net/browse/EMQX-12503?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by qzhuyan — DISMISSED** [review]

File: apps/emqx/src/emqx_frame.erl [hunk 1]
```diff
@@ -287,14 +287,23 @@ parse_connect(FrameBin, StrictMode) ->
...
     StrictMode
 ) ->
     case Reserved of
         0 -> ok;
         1 -> ?PARSE_ERR(reserved_connect_flag)
     end,
+    WillFlag = bool(WillFlagB),
+    WillRetain = bool(WillRetainB),
+    case {WillFlag, WillRetain, WillQoS > 0} of
+        %% [MQTT-3.1.2-15]
```
Comment (qzhuyan) [inline apps/emqx/src/emqx_frame.erl:301]:
```suggestion
        %% [MQTT-3.1.2-13]
```

File: apps/emqx/src/emqx_frame.erl [hunk 2]
```diff
@@ -287,14 +287,23 @@ parse_connect(FrameBin, StrictMode) ->
...
+        {false, true, _} -> ?PARSE_ERR(invalid_will_retain);
+        %% [MQTT-3.1.2-13]
```
Comment (qzhuyan) [inline apps/emqx/src/emqx_frame.erl:303]:
```suggestion
        %% [MQTT-3.1.2-11]
```

File: apps/emqx/src/emqx_frame.erl (hunk 1 above)
Comment (JimMoen) [reply apps/emqx/src/emqx_frame.erl:301]: [MQTT-v3.1.1 [MQTT-3.1.2-15]](https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#:~:text=If%20the%20Will%20Flag%20is%20set%20to%200%2C%20then%20the%20Will%20Retain%20Flag%20MUST%20be%20set%20to%200%20%5BMQTT%2D3.1.2%2D15%5D.)
[MQTT-v5.0 [MQTT-3.1.2-13]](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#:~:text=If%20the%20Will%20Flag%20is%20set%20to%200%2C%20then%20Will%20Retain%20MUST%20be%20set%20to%200%20%5BMQTT%2D3.1.2%2D13%5D)

File: apps/emqx/src/emqx_frame.erl (hunk 2 above)
Comment (JimMoen) [reply apps/emqx/src/emqx_frame.erl:303]: [MQTT-v3.1.1 [MQTT-3.1.2-13]](https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#:~:text=If%20the%20Will%20Flag%20is%20set%20to%200%2C%20then%20the%20Will%20QoS%20MUST%20be%20set%20to%200%C2%A0(0x00)%20%5BMQTT%2D3.1.2%2D13%5D.)
[MQTT-v5.0 [MQTT-3.1.2-11]](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#:~:text=If%20the%20Will%20Flag%20is%20set%20to%200%2C%20then%20the%20Will%20QoS%20MUST%20be%20set%20to%200%20(0x00)%20%5BMQTT%2D3.1.2%2D11%5D.)

thanks for catching. comments will be updated in later PR.

**Comment by JimMoen** [comment]
fix in v/e5.7.2
merge after v/e5.7.1 released

**Review by qzhuyan — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-2711)

**Pull request merged**